src/ripple/app/tx/impl/CreateTicket.cpp

```diff
@@ -23,105 +23,137 @@
 #include <ripple/protocol/Feature.h>
 #include <ripple/protocol/Indexes.h>
 #include <ripple/protocol/TxFlags.h>
+#include <ripple/protocol/st.h>
 
 namespace ripple {
 
-NotTEC
+std::pair<NotTEC, TxConsequences>
 CreateTicket::preflight (PreflightContext const& ctx)
 {
-    if (! ctx.rules.enabled(featureTickets))
-        return temDISABLED;
+    TxConsequences const conseq {ctx.tx, calculateSequencesConsumed (ctx.tx)};
+    if (! ctx.rules.enabled (featureTicketBatch))
+        return {temDISABLED, conseq};
 
     if (ctx.tx.getFlags() & tfUniversalMask)
-        return temINVALID_FLAG;
+        return {temINVALID_FLAG, conseq};
 
-    auto const ret = preflight1 (ctx);
-    if (!isTesSuccess (ret))
-        return ret;
+    if (std::uint32_t const count = {ctx.tx[sfCount]};
+                (count < minValidCount || count > maxValidCount))
+            return {temINVALID_COUNT, conseq};
 
-    if (ctx.tx.isFieldPresent (sfExpiration))
-    {
-        if (ctx.tx.getFieldU32 (sfExpiration) == 0)
-        {
-            JLOG(ctx.j.warn()) <<
-                "Malformed transaction: bad expiration";
-            return temBAD_EXPIRATION;
-        }
-    }
+    if (NotTEC const ret {preflight1 (ctx)}; ! isTesSuccess (ret))
+        return {ret, conseq};
+
+    return {preflight2 (ctx), conseq};
+}
+
+std::uint32_t
+CreateTicket::calculateSequencesConsumed(STTx const& tx)
+{
+    return tx[sfCount];
+}
+
+TER
+CreateTicket::preclaim(PreclaimContext const& ctx)
+{
+    auto const id = ctx.tx[sfAccount];
+    auto const sleAccountRoot = ctx.view.read(keylet::account(id));
+    if (! sleAccountRoot)
+        return terNO_ACCOUNT;
+
+    // Make sure the TicketCreate would not cause the account to own
+    // too many tickets.
+    std::uint32_t const curTicketCount =
+        {(*sleAccountRoot)[~sfTicketCount].value_or (0u)};
+    std::uint32_t const addedTickets = {ctx.tx[sfCount]};
```

Comment:
I think we should include @HowardHinnant in this discussion.

Reply:
I recommended this style.  But it _is_ just a guideline.  I have no intention of trying to enforce it.  However I'm happy to share my rationale.

Yes, `{}` for the narrowing catch.

```c++
uint32_t x = {y};
```

The `=` part is to catch at compile-time an accidental explicit conversion from `y` to `x`.  For example if `y` were to accidentally have the type `std::chrono::day`, then the above would not compile.  However this would compile:

```c++
uint32_t x{y};
```

The author of the type `day` recognized that implicit conversions from `day` to `unsigned` would be dangerous.  But nevertheless the conversion is still useful, so it was made explicit.  By having the client use implicit syntax, the client is saying:  I only want this to happen if it is a safe conversion (or no conversion at all).

With the syntax as coded, we don't have to check if `(*sleAccountRoot)[~sfTicketCount].value_or (0u)` returns a class type with an `explicit operator unsigned()`.  We have more confidence that it is returning an integral type.  And with the `{}`, an unsigned integral type that won't overflow when assigned to `uint32_t`.

Now as coded, closer inspection of `(*sleAccountRoot)[~sfTicketCount].value_or (0u)` makes it likely that this is returning `optional<unsigned>`.  But the point of the guideline is that as one gets used to the `integral x = {y}` syntax, one doesn't even have to inspect `y` to know it is a safe, non-narrowing initialization.

Reply:
Thanks @HowardHinnant I understand the rational. I'm still against doing this in isolation like this. I'm not against making this a recommended way of doing initialization from now on - tho we should have a discussion among the team first.

Also, I apologize if my comment about this being an "experiment" came across poorly.

Reply:
No it didn't come across poorly.  No apologies needed.  I was just trying to give background:  As in I didn't come up with this last week.  Been thinking about this and experimenting for quite a while.

About isolation:  Every marathon begins with one step. :-)  I think dedicating time to a sweeping change in style across all of rippled for any reason is not the best use of our limited time.  I recommend applying best practices to the code you're working on today.  This enables better practices to propagate more widely, and poor practices to die out, with little investment in time.

Reply:
@HowardHinnant @scottschurr I've thought about this, and I'm still against it. The benefit of this `int var = {foo()}` idiom over `int var{foo()}` is it protects against accidentally invoking an `explicit operator int()`. That is realistically never going to happen in this case.

If we are going to introduce this idiom, it would be good if the first place it was introduced had a realistic chance of catching an error - in some generic code say. If we find that place and think we want to change style from now forward, then we talk with the team and make the style change. As is, I still think the cost of making the code different from the rest of rippled outweighs the benefit here.

As usual, there's no "right" answer to this. I suspect I'm weighing code consistency more heavily than you two, and you two are trying to introduce a safer initialization mechanism. While I'm still a "vote no" for the code here, I also understand if it's kept as-is.

Reply:
This issue isn't going to impact my review one way or the other (sorry I'm running behind on that).  @scottschurr, sorry if I had you do a lot of work for nothing.  This looks like your call as to which way to resolve this, and however you do, I'm not loosing any sleep over it.